src/wrapper.cc

```diff
@@ -124,11 +125,14 @@ Dart_CObject CallbackArgBuilder(int num, ...) {
   Dart_CObject c_request_data;
   va_list valist;
   va_start(valist, num);
-  void *request_buffer = malloc(sizeof(uint64_t) * num);
+  // Initializing [request_buffer] with 0 will allow us to ensure padding in
+  // the higher 32 bits of 64 bit uint in case of systems with word size of 4
+  // bytes.
+  void *request_buffer = calloc(num, sizeof(uint64_t));
   uint64_t *buf = reinterpret_cast<uint64_t *>(request_buffer);
 
   for (int i = 0; i < num; i++) {
-    buf[i] = va_arg(valist, uint64_t);
+    buf[i] = va_arg(valist, intptr_t);

```

Comment:
That should be uintptr_t. intptr_t is signed.

`buf` is of type `uint64_t*` which means that an assignment (through `buf[i] =`) will always write the full 64 bits.

When assigning a `int32_t` (on 32 bit `intptr_t` equals `int32_t`), it will first be sign-extended to `int64_t` setting the upper 32 bits to 1 if the starting value was negative. And then it will convert `int64_t` to `uint64_t` (keeping the bits) before doing the write. So this will fail for memory addresses that have a 1 as the first byte on 32 bit.

More background: If you want to play around see https://godbolt.org/z/Kor3zK7TK and  https://godbolt.org/z/8fx9P9MWh and replace the C types and see what different assembly is generated, especially the arithmetic (right) shift instructions that do the sign extension. If you change the type to uint32_t the arithmetic shifts are replaced by simply writing 0 to the upper bits to zero them out.

Reply:
Oh! right. I totally missed the sign extension part.